modules/consensus/consensusset.go

```diff
@@ -139,12 +146,21 @@ func New(gateway modules.Gateway, persistDir string) (*ConsensusSet, error) {
 		}
 
 		// Register RPCs
-		gateway.RegisterRPC("SendBlocks", cs.rpcSendBlocks)
-		gateway.RegisterRPC("RelayBlock", cs.rpcRelayBlock) // COMPATv0.5.1
-		gateway.RegisterRPC("RelayHeader", cs.rpcRelayHeader)
-		gateway.RegisterRPC("SendBlk", cs.rpcSendBlk)
+		gateway.RegisterRPC("SendBlocks", cs.threadedRPCSendBlocks)
+		gateway.RegisterRPC("RelayBlock", cs.threadedRPCRelayBlock) // COMPATv0.5.1
+		gateway.RegisterRPC("RelayHeader", cs.threadedRPCRelayHeader)
+		gateway.RegisterRPC("SendBlk", cs.threadedRPCSendBlk)
 		gateway.RegisterConnectCall("SendBlocks", cs.threadedReceiveBlocks)
 
+		// Unregister RPCs on Close.
+		cs.tg.OnStop(func() {
+			cs.gateway.UnregisterRPC("SendBlocks")
+			cs.gateway.UnregisterRPC("RelayBlock") // COMPATv0.5.1
+			cs.gateway.UnregisterRPC("RelayHeader")
+			cs.gateway.UnregisterRPC("SendBlk")
+			cs.gateway.UnregisterConnectCall("SendBlocks")
+		})
+
```

Comment:
Interesting, and it seems like this would apply to many uses of `OnStop`. I think you could get around this like so:
```go
unregister := func() { /* blah blah blah */ }
err = cs.tg.OnStop(unregister)
if err != nil {
    // already stopped
    unregister()
}
```
But this raises the question: why not embed this logic directly in `OnStop`? After all, if the threadgroup is stopped, we want the function to execute! So I propose we modify `OnStop` like so:
```go
func (tg *ThreadGroup) OnStop(fn func()) {
	if tg.IsStopped() {
		fn() // call function immediately instead of returning an error
	} else {
		tg.mu.Lock()
		tg.stopFns = append(tg.stopFns, fn)
		tg.mu.Unlock()
	}
}
```